clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -1403,25 +1452,54 @@ public boolean isDone() {
     }
 
     /**
-     * A callback called when producer request is complete. It in turn calls user-supplied callback (if given) and
-     * notifies producer interceptors about the request completion.
+     * Callbacks that are called by the RecordAccumulator append functions:
+     *  - user callback
+     *  - interceptor callbacks
+     *  - partition callback
      */
-    private static class InterceptorCallback<K, V> implements Callback {
+    private class AppendCallbacks<K, V> implements RecordAccumulator.AppendCallbacks {
         private final Callback userCallback;
         private final ProducerInterceptors<K, V> interceptors;
-        private final TopicPartition tp;
+        private final ProducerRecord<K, V> record;
+        protected int partition = RecordMetadata.UNKNOWN_PARTITION;
 
-        private InterceptorCallback(Callback userCallback, ProducerInterceptors<K, V> interceptors, TopicPartition tp) {
+        private AppendCallbacks(Callback userCallback, ProducerInterceptors<K, V> interceptors, ProducerRecord<K, V> record) {
             this.userCallback = userCallback;
             this.interceptors = interceptors;
-            this.tp = tp;
+            this.record = record;
         }
 
+        @Override
         public void onCompletion(RecordMetadata metadata, Exception exception) {
-            metadata = metadata != null ? metadata : new RecordMetadata(tp, -1, -1, RecordBatch.NO_TIMESTAMP, -1, -1);
+            if (metadata == null) {
+                metadata = new RecordMetadata(topicPartition(), -1, -1, RecordBatch.NO_TIMESTAMP, -1, -1);
+            }
             this.interceptors.onAcknowledgement(metadata, exception);
             if (this.userCallback != null)
                 this.userCallback.onCompletion(metadata, exception);
         }
+
+        @Override
+        public void setPartition(int partition) {
+            assert partition != RecordMetadata.UNKNOWN_PARTITION;
+            this.partition = partition;
+
+            if (log.isTraceEnabled()) {
+                // Log the message here, because we don't know the partition before that.
+                log.trace("Attempting to append record {} with callback {} to topic {} partition {}", record, userCallback, record.topic(), partition);
+            }
+        }
+
+        public int getPartition() {
+            return partition;
+        }
+
+        public TopicPartition topicPartition() {
+            if (record == null)
```

Comment:
We have a bunch of code in send() that depends on record being not null. Perhaps it's better to assert non-null record early in send()?

Reply:
I've added it because there is a test that passes a null and expects some error handling.  From that perspective, adding an assert would contradict the expectation that there is defined and tested error handling (in my mind, an assert means the behavior is undefined and all bets are off).

Reply:
Hmm, by returning a null, we are breaking the contract that RecordMetadata.topic() and RecordMetadata.partition() is never null in producer callback and interceptor. Is that a new or existing test?

Reply:
This is an existing test https://github.com/apache/kafka/blob/trunk/core/src/test/scala/integration/kafka/api/PlaintextConsumerTest.scala#L1041
It passes in null and expects that the error goes through the interceptor's callback, but topicPartition() is called to supply an argument to the callback (which can handle null topicPartition) and it was throwing NPE.
I personally would just declare record as non-null and treat it as a bug in the application (undefined behavior), but as defined currently, it's a tested run-time error condition with some expected behavior.

Reply:
Thanks for the explanation. Since this is the existing behavior, we could keep the code as it is in the PR and improve it in the future.